cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/common/utils.go

```diff
@@ -52,52 +55,76 @@ func init() {
 //
 // miscellaneous utility funcs
 //
-func PrettyPrintObject(o runtime.Object) string {
+func PrettyPrint(o interface{}) string {
 	prettyBytes, err := json.MarshalIndent(o, "", "  ")
 	if err != nil {
 		return fmt.Sprintf("%v", o)
 	}
 	return string(prettyBytes)
 }
 
-func PrettyPrintMap(m map[string]interface{}) string {
-	prettyBytes, err := json.MarshalIndent(m, "", "  ")
+func FromUnstructured(unstructuredObj *unstructured.Unstructured, obj interface{}) error {
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredObj.Object, obj)
 	if err != nil {
-		return fmt.Sprintf("%v", m)
+		return status.Errorf(codes.Internal,
+			"failed to convert unstructured content for %s to a typed: %v",
+			PrettyPrint(unstructuredObj), err)
 	}
-	return string(prettyBytes)
+	return nil
 }
 
-// Confirm the state we are observing is for the current generation
-// returns true if object's status.observedGeneration == metadata.generation
-// false otherwise
-func CheckGeneration(unstructuredObj map[string]interface{}) bool {
-	observedGeneration, found, err := unstructured.NestedInt64(unstructuredObj, "status", "observedGeneration")
-	if err != nil || !found {
-		return false
+func ToUnstructured(obj interface{}) (*unstructured.Unstructured, error) {
+	unstructuredObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal,
+			"failed to convert typed content for %s to unstructured: %v",
+			PrettyPrint(obj), err)
 	}
-	generation, found, err := unstructured.NestedInt64(unstructuredObj, "metadata", "generation")
-	if err != nil || !found {
+	return &unstructured.Unstructured{Object: unstructuredObj}, nil
+}
+
+func CheckGeneration(obj ctrlclient.Object) bool {
+	generation := obj.GetGeneration()
+	var observedGeneration int64
+	if repo, ok := obj.(*sourcev1.HelmRepository); ok {
+		observedGeneration = repo.Status.ObservedGeneration
+	} else if rel, ok := obj.(*helmv2.HelmRelease); ok {
+		observedGeneration = rel.Status.ObservedGeneration
+	} else {
+		log.Errorf("Unsupported object type in CheckGeneration: %#v", obj)
 		return false
 	}
 	return generation == observedGeneration
 }
 
-func NamespacedName(unstructuredObj map[string]interface{}) (*types.NamespacedName, error) {
+func NamespacedName(obj ctrlclient.Object) (*types.NamespacedName, error) {
+	name := obj.GetName()
+	namespace := obj.GetNamespace()
+	if name != "" && namespace != "" {
+		return &types.NamespacedName{Name: name, Namespace: namespace}, nil
+	} else {
+		return nil,
+			status.Errorf(codes.Internal,
+				"required fields 'metadata.name' and/or 'metadata.namespace' not found on resource: %v",
+				PrettyPrint(obj))
+	}
+}
+
+func NamespacedNameForUnstructured(unstructuredObj map[string]interface{}) (*types.NamespacedName, error) {
```

Comment:
I wonder if using the built-in decoder in k8s `UniversalDecoder()` + `DecodeInto(...)`  could simplify the logic [1]. I mean, I feel doing things like `foo, found, err := unstructured.NestedString(unstructuredObj, "my", "path", "to, "foo")` is not the best approach, but I don't have a more precise suggestion atm.

[1] https://stackoverflow.com/a/63338697

Reply:
I wouldn't worry about it. This function is only called once in the entire code and even this 1 usage will go away (and the function foo), when I am done with the next step